polar-core/src/vm.rs

```diff
@@ -1889,25 +1969,185 @@ impl PolarVirtualMachine {
-        args: Vec<Term>,
-    ) -> PolarResult<QueryEvent> {
-        assert_eq!(args.len(), 2);
-        let item = &args[0];
-        let iterable = &args[1];
-        match (item.value(), iterable.value()) {
+        eval: F,
+        handle_unbound_left_var: bool,
+        handle_unbound_right_var: bool,
+    ) -> PolarResult<QueryEvent>
+    where
+        F: Fn(&mut Self, &Term) -> PolarResult<QueryEvent>,
+    {
+        let Operation { operator: op, args } = term.value().as_expression().unwrap();
+
+        let mut args = args.clone();
+        assert!(args.len() >= 2);
+        let left = &args[0];
+        let right = &args[1];
+
+        match (left.value(), right.value()) {
             (Value::Expression(_), _)
             | (_, Value::Expression(_))
             | (Value::RestVariable(_), _)
-            | (_, Value::RestVariable(_)) => unreachable!("invalid syntax"),
+            | (_, Value::RestVariable(_)) => {
+                panic!("invalid query");
+            }
+            _ => {}
+        };
+
+        if let Value::Variable(r) = right.value() {
+            // A variable on the right, ground on the left.
+            if let VariableState::Bound(x) = self.variable_state(r) {
+                args[1] = x;
+                self.push_goal(Goal::Query {
+                    term: term.clone_with_value(Value::Expression(Operation {
+                        operator: *op,
+                        args,
+                    })),
+                })?;
+                return Ok(QueryEvent::None);
+            } else if !handle_unbound_right_var && left.value().as_symbol().is_err() {
+                return eval(self, term);
+            }
+        }
+
+        if let Value::Variable(l) = left.value() {
+            // A variable on the left, ground on the right.
+            if let VariableState::Bound(x) = self.variable_state(l) {
+                args[0] = x;
+                self.push_goal(Goal::Query {
+                    term: term.clone_with_value(Value::Expression(Operation {
+                        operator: *op,
+                        args,
+                    })),
+                })?;
+                return Ok(QueryEvent::None);
+            } else if !handle_unbound_left_var && right.value().as_symbol().is_err() {
+                return eval(self, term);
+            }
+        }
+
+        if left.value().as_symbol().is_ok() || right.value().as_symbol().is_ok() {
+            self.add_constraint(term)?;
+            return Ok(QueryEvent::None);
+        }
+
+        eval(self, term)
+    }
+
+    /// Evaluate comparison operations.
+    fn comparison_op_helper(&mut self, term: &Term) -> PolarResult<QueryEvent> {
+        let Operation { operator: op, args } = term.value().as_expression().unwrap();
+
+        assert_eq!(args.len(), 2);
+        let left = &args[0];
+        let right = &args[1];
+
+        match (left.value(), right.value()) {
+            // TODO(ap): ExternalInstance on one side.
```

Comment:
Wouldn't this be handled the same as the `(External, External)` case?

Reply:
Yes, thank you. a7f9f561